data_source.go

```diff
@@ -39,23 +49,38 @@ type DataSource interface {
 	ConfigurePulseLengths(int, int) error
 	ConfigureProjectorsBases(int, mat.Dense, mat.Dense, string) error
 	ChangeTriggerState(*FullTriggerState) error
-	ConfigureMixFraction(int, float64) error
+	ConfigureMixFraction(*MixFractionObject) error
 	WriteControl(*WriteControlConfig) error
 	SetCoupling(CouplingStatus) error
 	SetExperimentStateLabel(string) error
 	ChannelsWithProjectors() []int
-	Wait() error
-	ProcessSegments() error
-	RunDoneAdd()
-	RunDoneDone()
+	ProcessSegments(*dataBlock) error
+	RunDoneActivate()
+	RunDoneDeactivate()
+	// RunDoneWait()
 	ShouldAutoRestart() bool
 }
 
-// Wait returns when the source run is done, aka the source is stopped
-func (ds *AnySource) Wait() error {
-	// fmt.Println("ds.Wait")
+// RunDoneActivate adds one to ds.runDone, this should only be called in Start
+func (ds *AnySource) RunDoneActivate() {
+	ds.runMutex.Lock()
+	defer ds.runMutex.Unlock()
+	ds.sourceState = Active
+	ds.runDone.Add(1)
+}
+
+// RunDoneDeactivate calls Done on ds.runDone, this should only be called in Start
+func (ds *AnySource) RunDoneDeactivate() {
+	ds.runMutex.Lock()
+	ds.sourceState = Inactive
+	ds.runDone.Done()
+	ds.runMutex.Unlock()
+}
+
+// RunDoneWait returns when the source run is done, i.e., the source is stopped
+func (ds *AnySource) RunDoneWait() {
```

Comment:
This appears to be commented out of the interface?

Reply:
Unneeded there. It's only called from the `AnySource.Stop` method.